community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py

```diff
@@ -54,20 +53,101 @@
 TOT_REQ_CNT = 1000
 _MAINTENANCE_SBATCH_SCRIPT_PATH = dirs.custom_scripts / "perform_maintenance.sh"
 
-NodeStatus = Enum(
-    "NodeStatus",
-    (
-        "orphan",
-        "power_down",
-        "preempted",
-        "restore",
-        "resume",
-        "terminated",
-        "unbacked",
-        "unchanged",
-        "unknown",
-    ),
-)
+class NodeAction(Protocol):
+    def apply(self, nodes:List[str]) -> None:
+        ...
+
+    def __hash__(self):
+        return hash(type(self))
+
+    def __eq__(self, other):
+        return type(self) == type(other)
+
+class NodeActionResume(NodeAction):
+    def apply(self, nodes:List[str]) -> None:
+        """resume nodes via scontrol"""
+        count = len(nodes)
+        hostlist = util.to_hostlist_fast(nodes)
+        log.info(f"{count} instances to resume ({hostlist})")
+        run(f"{lookup().scontrol} update nodename={hostlist} state=power_up")
+
+class NodeActionIdle(NodeAction):
+    def apply(self, nodes:List[str]) -> None:
+        """idle nodes"""
+        count = len(nodes)
+        hostlist = util.to_hostlist_fast(nodes)
+        log.info(f"{count} nodes to idle ({hostlist})")
+        run(f"{lookup().scontrol} update nodename={hostlist} state=resume")
+
+class NodeActionRestart(NodeAction):
+    def apply(self, nodes:List[str]) -> None:
+        """start instances for nodes"""
+        count = len(nodes)
+        hostlist = util.to_hostlist_fast(nodes)
+        log.info(f"{count} instances restarted ({hostlist})")
+        start_instances(nodes)
+
+class NodeActionDown(NodeAction):
+    def __init__(self, reason:str="Instance stopped/deleted"):
+        self.reason = reason
+    
+    def __hash__(self):
+        return hash(str(type(self))+self.reason)
```

Comment:
Instead of doing hash of string concat, do hash of tuple `hash((type(self), self.reason))`